File: share/android/src/main/java/com/kakaotam/rnsdk/share/KakaoSdkShareApiModule.kt
package com.kakaotam.rnsdk.share

import com.facebook.react.bridge.*
import com.kakao.sdk.share.ShareClient
import com.kakao.sdk.share.WebSharerClient
import com.kakao.sdk.share.model.ImageUploadResult
import com.kakao.sdk.share.model.SharingResult
import com.kakaotam.rnkakaosdkcommon.PromiseTransformer
import com.kakaotam.rnkakaosdkcommon.ReactToParam
import com.kakaotam.rnkakaosdkcommon.SdkInitializer
import com.kakaotam.rnsdk.share.model.*

class KakaoSdkShareApiModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "KakaoSdkShareApiModule"
    }

    @ReactMethod
    private fun isKakaoTalkSharingAvailable(promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        promise.resolve(ShareClient.instance.isKakaoTalkSharingAvailable(reactContext.currentActivity!!))
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun shareCustom(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ShareCustomParam::class.java)!!
        ShareClient.instance.shareCustom(
            reactContext.currentActivity!!,
            p.templateId,
            p.templateArgs,
            p.serverCallbackArgs
        ) f@{ sharingResult: SharingResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(sharingResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun shareDefault(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ShareDefaultParam::class.java)!!
        ShareClient.instance.shareDefault(
            reactContext.currentActivity!!,
            p.templatable!!,
            p.serverCallbackArgs
        ) f@{ sharingResult: SharingResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(sharingResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun scrapImage(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ScrapImageParam::class.java)!!
        ShareClient.instance.scrapImage(
            p.imageUrl,
            p.secureResource
        ) f@{ imageUploadResult: ImageUploadResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(imageUploadResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun shareScrap(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ShareScrapParam::class.java)!!
        ShareClient.instance.shareScrap(
            reactContext.currentActivity!!,
            p.requestUrl,
            p.templateId,
            p.templateArgs,
            p.serverCallbackArgs
        ) f@{ sharingResult: SharingResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(sharingResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun uploadImage(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, UploadImageParam::class.java)!!
        ShareClient.instance.uploadImage(
            p.image,
            p.secureResource
        ) f@{ imageUploadResult: ImageUploadResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(imageUploadResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun makeCustomUrl(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, MakeCustomUrlParam::class.java)!!
        promise.resolve(
            WebSharerClient.instance.makeCustomUrl(
                p.templateId, p.templateArgs, p.serverCallbackArgs
            ).toString()
        )
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun makeDefaultUrl(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, MakeDefaultUrlParam::class.java)!!
        promise.resolve(
            WebSharerClient.instance.makeDefaultUrl(
                p.templatable!!, p.serverCallbackArgs
            ).toString()
        )
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun makeScrapUrl(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, MakeScrapUrlParam::class.java)!!
        promise.resolve(
            WebSharerClient.instance.makeScrapUrl(
                p.requestUrl, p.templateId, p.templateArgs, p.serverCallbackArgs
            ).toString()
        )
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }
}

File: share/ios/KakaoSdkShareApiModule.swift
//
//  KakaoSdkShareApiModule.swift
//  react-native-kakao-sdk-share
//
//  Created by kakao-tam on 2022/05/06.
//

import Foundation

import RnKakaoSdkCommon
import KakaoSDKShare

@objc(KakaoSdkShareApiModule)
class KakaoSdkShareApiModule: NSObject {

    @objc(isKakaoTalkSharingAvailable:withReject:)
    func isKakaoTalkSharingAvailable(_ resolve:@escaping RCTPromiseResolveBlock,
                                     reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            resolve(ShareApi.isKakaoTalkSharingAvailable())
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(shareCustom:withResolve:withReject:)
    func shareCustom(_ param:NSDictionary,
                     resolve:@escaping RCTPromiseResolveBlock,
                     reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            let p = param.toParam(ShareCustomParam.self)!
            ShareApi.shared.shareCustom(templateId: p.templateId,
                                        templateArgs: p.templateArgs,
                                        serverCallbackArgs: p.serverCallbackArgs) {
                (sharingResult: SharingResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    if let sharingResult = sharingResult {
                        resolve(SharingResultWrapper(sharingResult).toJsonString());
                    } else {
                        resolve(nil)
                    }
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(shareDefault:withResolve:withReject:)
    func shareDefault(_ param:NSDictionary,
                      resolve:@escaping RCTPromiseResolveBlock,
                      reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            let p = param.toParam(ShareDefaultParam.self)!
            ShareApi.shared.shareDefault(templatable: p.templatable,
                                         serverCallbackArgs: p.serverCallbackArgs) {
                (sharingResult: SharingResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    if let sharingResult = sharingResult {
                        resolve(SharingResultWrapper(sharingResult).toJsonString());
                    } else {
                        resolve(nil)
                    }
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(scrapImage:withResolve:withReject:)
    func scrapImage(_ param:NSDictionary,
                    resolve:@escaping RCTPromiseResolveBlock,
                    reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            let p = param.toParam(ScrapImageParam.self)!
            ShareApi.shared.imageScrap(imageUrl: p.imageUrl,
                                       secureResource: p.secureResource) {
                (imageUploadResult: ImageUploadResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(imageUploadResult?.toJsonString());
                }
            }

        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(shareScrap:withResolve:withReject:)
    func shareScrap(_ param:NSDictionary,
                    resolve:@escaping RCTPromiseResolveBlock,
                    reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            let p = param.toParam(ShareScrapParam.self)!
            ShareApi.shared.shareScrap(requestUrl: p.requestUrl,
                                       templateId: p.templateId,
                                       templateArgs: p.templateArgs,
                                       serverCallbackArgs: p.serverCallbackArgs) {
                (sharingResult: SharingResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    if let sharingResult = sharingResult {
                        resolve(SharingResultWrapper(sharingResult).toJsonString());
                    } else {
                        resolve(nil)
                    }
                }
            }

        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(uploadImage:withResolve:withReject:)
    func uploadImage(_ param:NSDictionary,
                     resolve:@escaping RCTPromiseResolveBlock,
                     reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            let p = param.toParam(UploadImageParam.self)!
            ShareApi.shared.imageUpload(image: p.image,
                                        secureResource: p.secureResource) {
                (imageUploadResult: ImageUploadResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(imageUploadResult?.toJsonString());
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(makeCustomUrl:withResolve:withReject:)
    func makeCustomUrl(_ param:NSDictionary,
                       resolve:@escaping RCTPromiseResolveBlock,
                       reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            let p = param.toParam(MakeCustomUrlParam.self)!
            if let url = ShareApi.shared.makeCustomUrl(templateId: p.templateId,
                                                       templateArgs: p.templateArgs,
                                                       serverCallbackArgs: p.serverCallbackArgs) {
                resolve(url.absoluteString)
            } else {
                resolve(nil)
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(makeDefaultUrl:withResolve:withReject:)
    func makeDefaultUrl(_ param:NSDictionary,
                        resolve:@escaping RCTPromiseResolveBlock,
                        reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            let p = param.toParam(MakeDefaultUrlParam.self)!
            if let url = ShareApi.shared.makeDefaultUrl(
                templatable: p.templatable,
                serverCallbackArgs: p.serverCallbackArgs) {
                resolve(url.absoluteString)
            } else {
                resolve(nil)
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(makeScrapUrl:withResolve:withReject:)
    func makeScrapUrl(_ param:NSDictionary,
                      resolve:@escaping RCTPromiseResolveBlock,
                      reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = RnKakaoSdkObjC.tryCatch {
            RnKakaoSdkInitializer.shared.ensure()
            let p = param.toParam(MakeScrapUrlParam.self)!
            if let url = ShareApi.shared.makeScrapUrl(requestUrl: p.requestUrl, templateId: p.templateId, templateArgs: p.templateArgs,
                                                      serverCallbackArgs: p.serverCallbackArgs) {
                resolve(url.absoluteString)
            } else {
                resolve(nil)
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
}
